cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java

```diff
@@ -0,0 +1,167 @@
+ * Wrapper around the dataproc client that adheres to our configuration settings.
+ */
+public class DataProcClient implements AutoCloseable {
+  private final DataProcConf conf;
+  private final ClusterControllerClient client;
+
+  public static DataProcClient fromConf(DataProcConf conf) throws IOException {
+    ClusterControllerClient client = ClusterControllerClient.create(conf.getControllerSettings());
+    return new DataProcClient(conf, client);
+  }
+
+  private DataProcClient(DataProcConf conf, ClusterControllerClient client) {
+    this.conf = conf;
+    this.client = client;
+  }
+
+  public OperationSnapshot createCluster(String name) throws RetryableProvisionException {
+
+    // TODO: figure out how to set labels
+    try {
+      Cluster cluster = com.google.cloud.dataproc.v1.Cluster.newBuilder()
+        .setClusterName(name)
+        .setConfig(ClusterConfig.newBuilder()
+                     .setMasterConfig(InstanceGroupConfig.newBuilder()
+                                        .setNumInstances(conf.getMasterNumNodes())
+                                        .setMachineTypeUri(conf.getMasterMachineType())
+                                        .setDiskConfig(DiskConfig.newBuilder()
+                                                         .setBootDiskSizeGb(conf.getMasterDiskGB())
+                                                         .setNumLocalSsds(0)
+                                                         .build())
+                                        .build())
+                     .setWorkerConfig(InstanceGroupConfig.newBuilder()
+                                        .setNumInstances(conf.getWorkerNumNodes())
+                                        .setMachineTypeUri(conf.getWorkerMachineType())
+                                        .setDiskConfig(DiskConfig.newBuilder()
+                                                         .setBootDiskSizeGb(conf.getWorkerDiskGB())
+                                                         .setNumLocalSsds(0)
+                                                         .build())
+                                        .build())
+                     .setGceClusterConfig(GceClusterConfig.newBuilder()
+                                            .setSubnetworkUri("default")
+                                            .setZoneUri(conf.getZone())
+                                            .build())
+                     .build())
+        .build();
+
+      return client.createClusterAsync(conf.getProjectId(), conf.getRegion(), cluster)
+        .getInitialFuture().get(10, TimeUnit.SECONDS);
+    } catch (ApiException e) {
+      throw handleApiException(e);
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof ApiException) {
+        throw handleApiException((ApiException) cause);
+      }
+      throw Throwables.propagate(e);
+    }  catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```

Comment:
If you interrupt the current thread again, meaning the caller thread of this method would still get interrupted (e.g. on calling sleep inside RetryStrategy), is the what you wanted?

Reply:
I wasn't quite sure what to do, but now that I think about it some more, I think the right thing is not to catch it and let the caller handle it.